src/pymmcore_widgets/hcs/_util.py

```diff
@@ -0,0 +1,230 @@
+from __future__ import annotations
+
+import math
+from itertools import product
+from typing import TYPE_CHECKING
+
+import numpy as np
+from qtpy.QtCore import QRectF, Qt
+from qtpy.QtWidgets import (
+    QGraphicsScene,
+    QGraphicsView,
+    QWidget,
+)
+from useq import WellPlate
+
+from ._graphics_items import _WellGraphicsItem
+
+if TYPE_CHECKING:
+    from qtpy.QtGui import QBrush, QPen, QResizeEvent
+    from useq._grid import GridPosition
+
+    from ._graphics_items import Well
+
+import useq
+
+# add few coverslips to the known well plates
+useq.register_well_plates(
+    {
+        "coverslip-18mm": {
+            "rows": 1,
+            "columns": 1,
+            "well_spacing": 0.0,
+            "well_size": 18.0,
+            "circular_wells": False,
+            "name": "18mm coverslip",
+        },
+        "coverslip-22mm": {
+            "rows": 1,
+            "columns": 1,
+            "well_spacing": 0.0,
+            "well_size": 22.0,
+            "circular_wells": False,
+            "name": "22mm coverslip",
+        },
+    }
+)
+
+
+def custom_sort_key(item: str) -> tuple[int, int | str]:
+    """Sort well plate keys by number first, then by string."""
+    parts = item.split("-")
+    return (0, int(parts[0])) if parts[0].isdigit() else (1, item)  # type: ignore
+
+
+# sort the well plates by number first, then by string
+sorted_wells = sorted(useq.known_well_plate_keys(), key=custom_sort_key)
+
+# create a plate database with all known well plates
+PLATES: dict[str, WellPlate] = {}
+for key in sorted_wells:
+    plate = useq.WellPlate.from_str(key)
+    if isinstance(plate, WellPlate):
+        if not plate.name:
+            plate = plate.replace(name=key)
+        PLATES[key] = plate
+    elif isinstance(plate, dict):
+        PLATES[key] = WellPlate(**plate).replace(name=key)
+
+
+class _ResizingGraphicsView(QGraphicsView):
+    """A QGraphicsView that resizes the scene to fit the view."""
+
+    def __init__(self, scene: QGraphicsScene, parent: QWidget | None = None) -> None:
+        super().__init__(scene, parent)
+
+    def resizeEvent(self, event: QResizeEvent) -> None:
+        self.fitInView(self.sceneRect(), Qt.AspectRatioMode.KeepAspectRatio)
+
+
+def draw_plate(
```

Comment:
most of these seems be very similar to (and more verbose than) the `plot` function now in `useq`.  Perhaps try to mimic some of the patterns there, using things like `WellPlatePlan.all_well_positions` to do the heavy lifting on determining drawing coordinates?

Reply:
It is slightly different because I wanted to keep the ratio between well_size_x and well_size_y constant or the plate will look smaller or bigger depending on the size...